File: 2257-count-unguarded-cells-in-the-grid/2257-count-unguarded-cells-in-the-grid.rs
//  matrix cell indicators:
        //      0 - not guarded
        //      1 - guarded
        //      2 - guard
        //      3 - wall

use std::cmp::max;

impl Solution {
    pub fn count_unguarded(m: i32, n: i32, guards: Vec<Vec<i32>>, walls: Vec<Vec<i32>>) -> i32 {
        
        let (UNGUARDED, GUARDED, GUARD, WALL) = (0, 1, 2, 3);
        
        let mut cnt = (m * n) - guards.len() as i32 - walls.len() as i32;
        
        let mut matrix: Vec<Vec<i32>> = vec![vec![0; n as usize]; m as usize];
        
        for wall in walls{
            let i = wall[0] as usize;
            let j = wall[1]  as usize;
            
            matrix[i][j] = WALL;
        }
        
        for guard in guards.iter(){
            let i = guard[0]  as usize;
            let j = guard[1]  as usize;
            
            matrix[i][j] = GUARD;
        }
        
        let dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)];
        
        // For every guard change all unguarded cells 
        // in all four cardinal directions from it to
        // guarded cells.
        // Stop this process if it encounters another 
        // guard or a wall
        
        for guard in guards{
            for &(dx, dy) in dirs.iter(){
                let mut x = guard[0] + dx;
                let mut y = guard[1] + dy; 
                
                while x >= 0 && y >= 0 && x<m && y<n {
                    let (i,j) = (x as usize, y as usize);
                    
                    if matrix[i][j] == WALL || matrix[i][j] == GUARD{
                        break;
                    }
                    
                    if matrix[i][j] == UNGUARDED{
                        cnt -=1;
                    }
                    matrix[i][j] = GUARDED;
                    x += dx;
                    y += dy;
                } 
            }
            
        }
        
        cnt
    }
}

File: 2257-count-unguarded-cells-in-the-grid/2257-count-unguarded-cells-in-the-grid.py
# Each number in the grid will represent an object from 
# the problem statement
# 
# 0  -> Unguarded Cells
# 1  -> Guarded Cells
# 7  -> Guards
# 10 -> Walls

(UNGUARDED, GUARDED, GUARD, WALL) = (0, 1, 7, 10)

class Solution:
    def countUnguarded(self, m: int, n: int, guards: List[List[int]], walls: List[List[int]]) -> int:
        grid = [[0 for _ in range(0, n)] for _ in range(0, m)]
        
        cnt = ( m * n) - len(walls) - len(guards)
        
        for (row, col) in walls:
            grid[row][col] = WALL
            
        for (row, col) in guards:
            grid[row][col] = GUARD
        
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        
        for (row, col) in guards:
            
            for dx, dy in dirs:
                x = row + dx
                y = col + dy
                
                while ( 
                    x>=0 and 
                    x <m and 
                    y >= 0 and 
                    y< n and 
                    grid[x][y] != WALL and 
                    grid[x][y] != GUARD
                ):
                    if grid[x][y] == UNGUARDED:
                        cnt-=1
                    grid[x][y] = GUARDED
                    x = x + dx
                    y = y + dy

        return cnt  
            
        